In [2]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

dataset = pd.read_csv('churndataset.csv')
x = dataset.iloc[:,3:-1]
y = dataset['Exited']
dataset['Exited'].value_counts()
dataset

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler

x_onehot = pd.get_dummies(x, drop_first = True)
x_train, x_test, y_train, y_test = train_test_split(x_onehot,y,test_size=0.2,random_state=42, stratify=y)
x_train.describe()

scaler = StandardScaler()
x_train = scaler.fit_transform(x_train)
x_test = scaler.transform(x_test)

import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Flatten, Dense, InputLayer

from tensorflow.keras.optimizers import Adam

input_shape = x_train.shape[1:]
class_weight = y.value_counts().to_dict()
total_sample = sum(class_weight.values())
class_weight = [total_sample/(len(class_weight)*class_weight[0]), total_sample/(len(class_weight)*class_weight[1])]
class_weight

model = Sequential()
model.add(InputLayer(input_shape))
model.add(Dense(128, activation = 'relu'))
model.add(Dense(128, activation = 'relu'))
model.add(Dense(1, activation = 'sigmoid'))

model.summary()

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense (Dense)               (None, 128)               1536      
                                                                 
 dense_1 (Dense)             (None, 128)               16512     
                                                                 
 dense_2 (Dense)             (None, 1)                 129       
                                                                 
Total params: 18177 (71.00 KB)
Trainable params: 18177 (71.00 KB)
Non-trainable params: 0 (0.00 Byte)
_________________________________________________________________


## Layer 1: Dense Layer

**dense (Dense)               
Output Shape: (None, 128)   
Param #: 1,536**

- Input Shape: (11,) → This layer expects 11 input features.
- Neurons (Units): 128 neurons in this layer.
- Activation Function: If not specified, it uses the default activation (usually ReLU).

#### Parameter Calculation (Params)

The number of parameters in a Dense layer is calculated as:

- Params = (Input Features + 1) x neurons
- i.e: (11 + 1) x 128 = 1,536
- The extra +1 is for the bias term.

## Layer 2: Dense Layer

**dense_1 (Dense)            
Output Shape: (None, 128)  
Param #: 16,512**

- This layer takes 128 inputs from the previous layer and outputs 128 neurons.

#### Parameter Calculation

- (128+1) x 128 = 16,512

## Layer 3: Dense Layer (Output Layer)

**dense_2 (Dense)            
Output Shape: (None, 1)    
Param #: 129**

- This is the output layer, with 1 neuron. It’s likely a regression model (predicting a continuous value) or a binary classification if using a sigmoid activation.

#### Parameter Calculation

- (128 + 1) x 1 = 129

## Total Parameters Calculation

- Adding up all the parameters:

    - 1536 + 16512 + 129 = 18177 -> total trainable parameters
 
## Summary:
- Input Shape: 11 columns (features).
- First Dense Layer: 128 neurons, with 1,536 trainable parameters.
- Second Dense Layer: 128 neurons, with 16,512 trainable parameters.
- Output Layer: 1 neuron, with 129 trainable parameters.
- Total Trainable Params: 18,177 parameters.